Re-introduce "input.enablejoystick" setting

garbear opened the pull request:
## Description

The `input.enablejoystick` setting was removed in the joystick subsystem introduced in https://github.com/xbmc/xbmc/pull/8807. Instead of this setting, controller input could be disabled from the GUI by disabling the peripheral.joystick add-on.

## Motivation and Context

The reason for re-adding this setting is due to Kodi's interaction with the window manager. On systems without a window manager, Kodi won't receive an event when the main app loses focus. By introducing this setting, an API is provided to clients so they can appropriately enable controller input based on external knowledge.

Discussion: https://forum.kodi.tv/showthread.php?tid=352475

Fixes https://github.com/garbear/xbmc/issues/105

Fixes https://trac.kodi.tv/ticket/16464

## How Has This Been Tested?

Test builds: https://github.com/garbear/xbmc/releases/tag/retroplayer-18.6-20200320

(tested on Leia for convenience)

#### Disabling controllers

Ran the command:

```
curl -v --data-binary '{"jsonrpc": "2.0", "method": "Settings.SetSettingValue", "params": {"setting": "input.enablejoystick", "value": false}, "id": 1}' -H 'content-type: application/json;' http://localhost:8080/jsonrpc
```

Result in GUI:

![Controller disabled](https://user-images.githubusercontent.com/531482/77220804-e686fe00-6b00-11ea-980f-6a829f3de152.png)

Effect: Controller disabled

#### Enabling controllers

Ran the command:

```
curl -v --data-binary '{"jsonrpc": "2.0", "method": "Settings.SetSettingValue", "params": {"setting": "input.enablejoystick", "value": true}, "id": 1}' -H 'content-type: application/json;' http://localhost:8080/jsonrpc
```

Result in GUI:
![Controller enabled 2](https://user-images.githubusercontent.com/531482/77220829-3bc30f80-6b01-11ea-9de8-f07e104edaec.png)

Effect: Controller enabled

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

lrusak commented on xbmc/settings/Settings.cpp:
```diff
@@ -361,6 +361,7 @@ const std::string CSettings::SETTING_AUDIOOUTPUT_VOLUMESTEPS = "audiooutput.volu
 const std::string CSettings::SETTING_INPUT_PERIPHERALS = "input.peripherals";
 const std::string CSettings::SETTING_INPUT_PERIPHERALLIBRARIES = "input.peripherallibraries";
 const std::string CSettings::SETTING_INPUT_ENABLEMOUSE = "input.enablemouse";
+const std::string CSettings::SETTING_INPUT_ENABLE_CONTROLLER = "input.enablejoystick";
```
Would it make more sense to make this specific to where it's used rather than generally here?

garbear commented:
@lrusak updated

garbear commented:
Uploaded test builds here: https://github.com/garbear/xbmc/releases/tag/retroplayer-18.6-20200320

lrusak commented on xbmc/input/InputManager.h:
```diff
@@ -289,6 +294,12 @@ class CInputManager : public ISettingCallback,
+  bool m_enableController = true;
+
+  // Settings
+  static const std::string SETTING_INPUT_ENABLE_CONTROLLER;
```
Sorry to bother again. Does this have to be a member if it's only used in this class? I find defining these in an anonymous namespace to be nice and tidy. It's totally up to you though as I'm fine either way :)

MilhouseVH commented:
@garbear I'll include this in LibreELEC test builds starting with `#0321`

phunkyfish commented:
Anything to report on this?

garbear commented:
I'll see what they have to say over in https://github.com/garbear/xbmc/issues/105

garbear committed:
[peripherals] Style changes and parameter grouping in PeripheralJoystick.h

garbear committed:
[peripherals] Re-introduce "input.enablejoystick" setting

This setting was removed in the joystick subsystem introduced in Kodi 17
Krypton. Instead, controller input could be disabled from the GUI by
disabling the peripheral.joystick add-on.

The reason for re-adding this setting is due to Kodi's interaction with the
window manager. On systems without a window manager, Kodi won't receive an
event when the main app loses focus. By introducing this setting, an API is
provided to clients so they can appropriately control controller input based
on external knowledge.

garbear committed:
[strings] Reword re-introduced controller setting for consistency in the UI

This setting was removed in Kodi 17 Krypton, so both of these were dead
strings in 17 and 18.

In the v17 joystick subsystem, the European-preferred verbage of "joystick"
was changed to "controller" for a more general audience. Now that we're
re-introducing the setting strings, let's use "controller" for concistency.

garbear committed:
[strings] Remove unused joystick strings

garbear committed:
[settings] Make enable mouse/controller advanced settings

These options can kill Kodi if no alternative input method is available, so
let's hide them by default.

garbear commented:
I fixed the setting being ignored during gameplay. I also rebased on master and fixed the merge conflict.

Finally, I changed "Enable mouse/touch" and "Enable controller" settings to advanced settings. Both of these can kill Kodi input if there is no alternative input method, so we should probably hide them by default.

garbear commented:
jenkins build this please

garbear commented:
jenkins build this please

Rechi commented on xbmc/input/InputManager.cpp:
```diff
@@ -89,6 +92,8 @@ void CInputManager::InitializeInputs()
   m_Mouse.Initialize();
   m_Mouse.SetEnabled(CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(CSettings::SETTING_INPUT_ENABLEMOUSE));
+
+  m_enableController = CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(SETTING_INPUT_ENABLE_CONTROLLER);
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/peripherals/devices/PeripheralJoystick.cpp:
```diff
@@ -203,6 +207,14 @@ bool CPeripheralJoystick::OnButtonMotion(unsigned int buttonIndex, bool bPressed
+  // Check GUI setting and send button release if controllers are disabled
+  if (!m_manager.GetInputManager().IsControllerEnabled())
+  {
+    for (std::vector<DriverHandler>::iterator it = m_driverHandlers.begin(); it != m_driverHandlers.end(); ++it)
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/peripherals/devices/PeripheralJoystick.cpp:
```diff
@@ -246,6 +262,14 @@ bool CPeripheralJoystick::OnHatMotion(unsigned int hatIndex, HAT_STATE state)
+  // Check GUI setting and send hat unpressed if controllers are disabled
+  if (!m_manager.GetInputManager().IsControllerEnabled())
+  {
+    for (std::vector<DriverHandler>::iterator it = m_driverHandlers.begin(); it != m_driverHandlers.end(); ++it)
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/peripherals/devices/PeripheralJoystick.cpp:
```diff
@@ -294,6 +318,14 @@ bool CPeripheralJoystick::OnAxisMotion(unsigned int axisIndex, float position)
+  // Check GUI setting and send analog axis centered if controllers are disabled
+  if (!m_manager.GetInputManager().IsControllerEnabled())
+  {
+    for (std::vector<DriverHandler>::iterator it = m_driverHandlers.begin(); it != m_driverHandlers.end(); ++it)
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi reviewed (commented):
Please follow the [current code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

garbear commented:
Fixed formatting, but jenkins failed after 10 seconds? jenkins build this please

garbear commented:
All better. Win32 failure unrelated. Test builds uploaded to: https://github.com/garbear/xbmc/releases/tag/retroplayer-18.6-20200516

garbear commented:
Marked https://trac.kodi.tv/ticket/16464 as solved. Gotta love the 5 year bug fix :)
